### PR TITLE
update icewind/streams to 0.4.1

### PR DESCRIPTION
**Pull request by icewind1991**
Requires https://github.com/owncloud/3rdparty/pull/278

Contains a fix for owncloud/core#23781

cc @owncloud/filesystem

### CODE REVIEW

**Comment by PVince81** [comment]
A bit risky to update libs after the freeze.

Also this would require regression testing of upload/download for a lot of external storages.

@owncloud/qa not sure if those have been (regression) tested already ?

**Comment by SergioBertolinSG** [comment]
Not really, @jvillafanez started some automated tests for external storage but not sure about the status.

We can do regression testing on external storages only for upload/download instead of the full test plan (which requires much time).

**Comment by icewind1991** [comment]
> A bit risky to update libs after the freeze.

No difference between updating a lib and updating core if it's all my code in the first place :smile: 

imo the change is safe and both the lib and it's usages in oC are covered with unit tests

**Comment by jvillafanez** [comment]
> Not really, @jvillafanez started some automated tests for external storage but not sure about the status.

Work in progress in https://github.com/owncloud/automated-tests/tree/upgrade_testing_external_storage I think I had something working for local storage, but there is still work to be done for the rest of the storages

**Comment by PVince81** [comment]
Okay, so I guess we can then merge this :+1: 

Second review please @owncloud/filesystem

**Comment by PVince81** [comment]
@georgehrke @nickvergessen @rullzer (also review 3rdparty one, it's not merged yet)

**Comment by PVince81** [comment]
Argh, this doesn't help fixing https://github.com/owncloud/core/issues/23781 because there are two instances of icewind/streams... the files_external one needs updating too @icewind1991

**Comment by icewind1991** [comment]
also updated the files_external version

File: apps/files_external/3rdparty/composer/autoload_static.php [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        'Icewind\\Streams\\Tests\\DirectoryWrapper' => __DIR__ . '/..' . '/icewind/streams/tests/DirectoryWrapper.php',
+        'Icewind\\Streams\\Tests\\DirectoryWrapperDummy' => __DIR__ . '/..' . '/icewind/streams/tests/DirectoryWrapper.php',
+        'Icewind\\Streams\\Tests\\DirectoryWrapperNull' => __DIR__ . '/..' . '/icewind/streams/tests/DirectoryWrapper.php',
+        'Icewind\\Streams\\Tests\\FailWrapper' => __DIR__ . '/..' . '/icewind/streams/tests/RetryWrapper.php',
+        'Icewind\\Streams\\Tests\\IteratorDirectory' => __DIR__ . '/..' . '/icewind/streams/tests/IteratorDirectory.php',
+        'Icewind\\Streams\\Tests\\NullWrapper' => __DIR__ . '/..' . '/icewind/streams/tests/NullWrapper.php',
+        'Icewind\\Streams\\Tests\\PartialWrapper' => __DIR__ . '/..' . '/icewind/streams/tests/RetryWrapper.php',
+        'Icewind\\Streams\\Tests\\RetryWrapper' => __DIR__ . '/..' . '/icewind/streams/tests/RetryWrapper.php',
+        'Icewind\\Streams\\Tests\\SeekableWrapper' => __DIR__ . '/..' . '/icewind/streams/tests/SeekableWrapper.php',
+        'Icewind\\Streams\\Tests\\UrlCallBack' => __DIR__ . '/..' . '/icewind/streams/tests/UrlCallBack.php',
```
Comment (PVince81) [inline apps/files_external/3rdparty/composer/autoload_static.php]: should the tests be removed here too?

**Comment by PVince81** [comment]
Works now.

File: apps/files_external/3rdparty/composer/autoload_static.php (hunk 1 above)
Comment (icewind1991) [inline apps/files_external/3rdparty/composer/autoload_static.php]: done

File: apps/files_external/3rdparty/composer/autoload_static.php [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        'Icewind\\SMB\\NativeServer' => __DIR__ . '/..' . '/icewind/smb/src/NativeServer.php',
+        'Icewind\\SMB\\NativeShare' => __DIR__ . '/..' . '/icewind/smb/src/NativeShare.php',
+        'Icewind\\SMB\\NativeState' => __DIR__ . '/..' . '/icewind/smb/src/NativeState.php',
+        'Icewind\\SMB\\NativeStream' => __DIR__ . '/..' . '/icewind/smb/src/NativeStream.php',
+        'Icewind\\SMB\\Parser' => __DIR__ . '/..' . '/icewind/smb/src/Parser.php',
+        'Icewind\\SMB\\RawConnection' => __DIR__ . '/..' . '/icewind/smb/src/RawConnection.php',
+        'Icewind\\SMB\\Server' => __DIR__ . '/..' . '/icewind/smb/src/Server.php',
+        'Icewind\\SMB\\Share' => __DIR__ . '/..' . '/icewind/smb/src/Share.php',
+        'Icewind\\SMB\\System' => __DIR__ . '/..' . '/icewind/smb/src/System.php',
+        'Icewind\\SMB\\Test\\AbstractShare' => __DIR__ . '/..' . '/icewind/smb/tests/AbstractShare.php',
```
Comment (PVince81) [inline apps/files_external/3rdparty/composer/autoload_static.php]: more tests to remove ?

Please be thourough

Comment (icewind1991) [inline apps/files_external/3rdparty/composer/autoload_static.php]: done

File: apps/files_external/3rdparty/composer/autoload_static.php [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+// autoload_static.php @generated by Composer
+
+namespace Composer\Autoload;
+
+class ComposerStaticInit98fe9b281934250b3a93f69a5ce843b3
+{
+    public static $prefixLengthsPsr4 = array (
+        'I' => 
+        array (
+            'Icewind\\Streams\\Tests\\' => 22,
```
Comment (PVince81) [inline apps/files_external/3rdparty/composer/autoload_static.php]: more code related to tests?

File: apps/files_external/3rdparty/composer/installed.json [hunk 4]
```diff
@@ -40,38 +41,37 @@
...
-        "time": "2016-04-26 13:26:39",
+        "time": "2016-06-02 14:37:52",
         "type": "library",
-        "installation-source": "source",
+        "installation-source": "dist",
         "autoload": {
             "psr-4": {
-                "Icewind\\SMB\\": "src/",
-                "Icewind\\SMB\\Test\\": "tests/"
+                "Icewind\\Streams\\Tests\\": "tests/",
```
Comment (PVince81) [inline apps/files_external/3rdparty/composer/installed.json]: same here

**Comment by PVince81** [comment]
Does composer has a way to separate tests from source ? It seems tedious to have to manually remove stuff.

I'm fine with the current state of this PR :+1:

**Comment by nickvergessen** [comment]
:+1: apart from the test stuff in the autoloader and json files

File: apps/files_external/3rdparty/composer/autoload_static.php [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            'Icewind\\Streams\\' => 16,
+            'Icewind\\SMB\\Test\\' => 17,
+            'Icewind\\SMB\\' => 12,
+        ),
+    );
+
+    public static $prefixDirsPsr4 = array (
+        'Icewind\\Streams\\Tests\\' => 
+        array (
+            0 => __DIR__ . '/..' . '/icewind/streams/tests',
```
Comment (nickvergessen) [inline apps/files_external/3rdparty/composer/autoload_static.php]:
```
Syntax error found in 1 file

------------------------------------------------------------

Parse error: ./apps/files_external/3rdparty/composer/autoload_static.php:22

    20|         'Icewind\\Streams\\Tests\\' => 

    21|         array (

  > 22|             0 => __DIR__ . '/..' . '/icewind/streams/tests',

    23|         ),

    24|         'Icewind\\Streams\\' => 

Unexpected '.', expecting ')'
```

Comment (icewind1991) [inline apps/files_external/3rdparty/composer/autoload_static.php]: turns out newer composer versions versions use constant expressions which are only supported by php >=5.6

Comment (PVince81) [inline apps/files_external/3rdparty/composer/autoload_static.php]: so can you fix it in a way that works for all PHP versions ? I remember you have a nice docker thingy for testing multiple versions so please use that and confirm that external storage + SFTP upload to read-only folder works in these versions. Thanks

**Comment by icewind1991** [comment]
re-created the autoloader with an older composer version

**Comment by PVince81** [comment]
Retested, works :+1: 

I have merged the 3rdparty PR and adjusted this PR to point to the latest 3rdparty master.

@nickvergessen second review ?

File: apps/files_external/3rdparty/composer/installed.json (hunk 4 above)
Comment (nickvergessen) [inline apps/files_external/3rdparty/composer/installed.json]: this is still open?

**Comment by nickvergessen** [comment]
anyway, since the code is only moved, fine by me :+1:

**Pull request merged**